### PR TITLE
Allow attributes on record fields

### PR DESCRIPTION
**Pull request by mateuszbujalski**
This change extends the record syntax to allow decorating fields with attributes. 

I still need to write a test where I query the attribute from a tactic to make sure it works correctly.

### CODE REVIEW

File: src/parser/FStar.Parser.ToDocument.fs [hunk 1]
```diff
@@ -973,7 +973,7 @@ and p_aqual = function
     in
     str "#[" ^^ p_term false false t ^^ str "]" ^^ break1
   | Meta (Arg_qualifier_meta_attr t) ->
-    str "#[@" ^^ p_term false false t ^^ str "]" ^^ break1
+    str "#[@@" ^^ p_term false false t ^^ str "]" ^^ break1
```
Comment (mateuszbujalski) [inline src/parser/FStar.Parser.ToDocument.fs:976]: I think the "#[@" syntax is deprecated? Is that correct?

File: src/parser/parse.fsy [hunk 2]
```diff
@@ -223,49 +224,49 @@ option___anonymous_1_:
 let x =                                                (t) in
     ( Some x )}
 
-option___anonymous_2_:
+option___anonymous_10_:
+  
+    {    ( None )}
+| LBRACE noSeqTerm RBRACE
```
Comment (mateuszbujalski) [inline src/parser/parse.fsy:230]: This doesn't seem related to my change. Not sure why this was generated.

File: src/syntax/FStar.Syntax.Resugar.fs [hunk 3]
```diff
@@ -1190,7 +1190,17 @@ let resugar_typ env datacon_ses se : sigelts * A.tycon =
...
               begin match (SS.compress term).n with
                 | Tm_arrow(bs, _) ->
-                  let mfields = bs |> List.map (fun (b, qual) -> (bv_as_unique_ident b, resugar_term' env b.sort)) in
+                  let mfields = 
+                    bs 
+                    |> List.map (fun (b, qual) -> 
+                        let q = 
+                            // TODO: There's a 'FIXME' note on resugar_arg_qual that I don't fully understand. It doesn't seem applicable in this case as the 
+                            //       syntax for records only support 'Arg_qualifier_meta_attr' and the only problematic case seems to be 'inaccessible implicit argument of a data constructor'
+                            match resugar_arg_qual env qual with
```
Comment (mateuszbujalski) [inline src/syntax/FStar.Syntax.Resugar.fs:1199]: @nikswamy This is my main question. The only supported qual for record field is an attribute, and looking at how 'resugar_arg_qual' works, it should always return 'Some'. Do you think this code is fine? Should I at least leave some comment here?

File: src/parser/FStar.Parser.ToDocument.fs (hunk 1 above)
Comment (nikswamy) [reply src/parser/FStar.Parser.ToDocument.fs:976]: Agreed, good to fix this. thanks.

**Comment by mateuszbujalski** [comment]
@nikswamy I've extended the test and querying seems to work fine. I consider this one done and ready for review :)

File: tests/micro-benchmarks/RecordFieldAttributes.fst [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+module T = FStar.Tactics
+
+type description (d : string) = ()
+
+type r = 
+    {
+        #[@@ (description "This is a number") ] field1 : int;
+        #[@@ (description "This is a string") ] field2 : string;
+    }
+
```
Comment (nikswamy) [inline tests/micro-benchmarks/RecordFieldAttributes.fst:12]: Sadly, there's one aspect of it that I neglected to notice earlier. These attributes on binders are unfortunately coupled with implicit arguments (i.e., the preceding `#`) makes both these fields implicits arguments to the record constructor.

That means that the code below fails to typecheck

```
@@expect_failure]
let test (x:int) (y:string) = { field1 = x; field2 = y }
```

And you instead have to write:

```
let test (x:int) (y:string) : r = Mkr #x #y
```

I think we should decouple attributes on binders with implicit arguments. i.e., you should be able to associate an attribute with a binder whether it is implicit or not. E.g., 

```
([@@ attr1] x : t) -> (#[@@attr2] y:t) -> ... 
```

The first being an explicit and the second an implicit arg.

@aseemr @R1kM  ... wdyt? This should be fine for Steel too, right?

Comment (R1kM) [reply tests/micro-benchmarks/RecordFieldAttributes.fst:12]: I think it should be fine for Steel indeed. So far, I don't think we needed the framing attribute on explicit arguments, but this would be good to have for the future

Comment (mateuszbujalski) [reply tests/micro-benchmarks/RecordFieldAttributes.fst:12]: Right, I completely missed this.

"I think we should decouple attributes on binders with implicit arguments." 
@nikswamy Do I understand correctly that 'Meta' aqual cases are currently assumed to be implicit? And I should extend the aqual with some new case which could be used for an explicit argument? Plus change the rule so that when HASH is missing, we go with the new case?

Comment (nikswamy) [reply tests/micro-benchmarks/RecordFieldAttributes.fst:12]: Yes, that's right.

**Comment by mateuszbujalski** [comment]
@nikswamy @aseemr I've updated the PR after the refactoring done by Aseem. I wrote some simple test for records and now it works both for implicit and explicit fields. I'll add some more tests for various cases where attributes are now possible. 

Unfortunately, I wasn't able to change the parser without increasing the number of shift/reduce conflicts around LBRACK_AT_AT (the total number went up from 6 to 11 - I've temporarily hacked the makefile to accept this to see if CI will be happy with the rest of the changes). What's currently pushed there seems to work but it's probably best if it's reverted and reworked. It would be great if you could take a look :)
